FLT/DedekindDomain/AdicValuation.lean

```diff
@@ -0,0 +1,288 @@
+    WithVal, HeightOneSpectrum.adicValued_apply,
+    HeightOneSpectrum.valuation_of_algebraMap] at hge
+  have hdz : (algebraMap A (WithVal (v.valuation K)) d) ≠ 0 :=
+    IsLocalization.to_map_ne_zero_of_mem_nonZeroDivisors _ (fun _ ↦ id) hd
+  -- Find a suitable `k` for the bound in `exists_adicValued_mul_sub_le`
+  have hv : Valued.v ((algebraMap A (WithVal (v.valuation K)) d)) ≠ 0 := by
+    rw [Valuation.ne_zero_iff]
+    exact hdz
+  let hu : Valued.v ((algebraMap A (WithVal (v.valuation K)) d)) * γ.val ≠ 0 := by
+    rw [mul_ne_zero_iff]
+    exact ⟨hv, γ.ne_zero⟩
+  obtain ⟨k, hku, hkv⟩ := exists_ofAdd_lt_and_ofAdd_lt hu hv
+  simp only [WithVal, HeightOneSpectrum.adicValued_apply,
+    HeightOneSpectrum.valuation_of_algebraMap, HeightOneSpectrum.intValuation_apply] at hkv
+  -- Now can apply `exists_adicValued_mul_sub_le` to get the approximation of x.
+  obtain ⟨a, hval⟩ := exists_adicValued_mul_sub_le v (nonZeroDivisors.ne_zero hd) hkv.le hge
+  use a
+  rw [← eq_div_iff_mul_eq hdz] at hnd
+  rw [← UniformSpace.Completion.coe_sub,
+      HeightOneSpectrum.valuedAdicCompletion_eq_valuation',
+      hnd, sub_div' hdz, map_div₀]
+  unfold WithVal at hdz ⊢
+  rw [← Valuation.pos_iff (HeightOneSpectrum.valuation K v)] at hdz
+  rw [← map_mul, ← map_sub, div_lt_iff₀' hdz, HeightOneSpectrum.valuation_of_algebraMap,
+      HeightOneSpectrum.intValuation_apply]
+  apply lt_of_le_of_lt hval hku
+
+/-- The closure of `A` in `K_v` is `𝒪_v`. -/
+theorem closureAlgebraMapIntegers_eq_integers :
+    closure (algebraMap A (v.adicCompletion K)).range =
+    SetLike.coe (v.adicCompletionIntegers K) := by
+  apply subset_antisymm
+  . apply closure_minimal _ Valued.valuationSubring_isClosed
+    rintro b ⟨a, rfl⟩
+    exact HeightOneSpectrum.coe_mem_adicCompletionIntegers v a
+  -- Show `𝒪_v ⊆ closure A` from `𝒪_v ⊆ closure O_[K]` and `closure O_[K] ⊆ closure A`
+  let f := fun (k : WithVal (v.valuation K)) => (k : v.adicCompletion K)
+  suffices h : closure (f '' (f ⁻¹' (HeightOneSpectrum.adicCompletionIntegers K v))) ⊆
+      closure (algebraMap A (HeightOneSpectrum.adicCompletion K v)).range by
+    apply Set.Subset.trans _ h
+    exact DenseRange.subset_closure_image_preimage_of_isOpen
+      UniformSpace.Completion.denseRange_coe (Valued.valuationSubring_isOpen _)
+  -- Unfold the topological definitions until we get the result from the previous lemma
+  apply closure_minimal _ isClosed_closure
+  rintro k ⟨x, hx, rfl⟩
+  unfold f at hx
+  rw [Set.mem_preimage, SetLike.mem_coe, HeightOneSpectrum.mem_adicCompletionIntegers,
+      Valued.valuedCompletion_apply, HeightOneSpectrum.adicValued_apply'] at hx
+  rw [mem_closure_iff_nhds_zero]
+  intro U hU
+  rw [Valued.mem_nhds] at hU
+  obtain ⟨γ, hγ⟩ := hU
+  obtain ⟨a, ha⟩ := exists_adicValued_sub_lt_of_adicValued_le_one K v γ hx
+  use algebraMap A K a
+  constructor
+  . use a
+    rfl
+  apply hγ
+  simpa
+
+/-- A is dense in `𝒪_v`. -/
+theorem denseRange_of_integerAlgebraMap :
```

Comment:
Getting this into mathlib would be a very natural project.

Reply:
There's a lot of mathlib projects I'm think about after this.